### ISSUE TITLE
Not redirecting to Salesforce login page while opening the app

### ISSUE THREAD

**Issue by aravindtrue** [forcedotcom/SalesforceMobileSDK-Android#1019]
forcedroid version 4.0.0
cordova version 5.4.1
phonegap version 5.3.9
cordova platform version android 5.0.0
mac OS X Yosemite 10.10.5

I followed the below steps to create a new app,

```
$ forcedroid create
   hybrid_local
   myApp
   /Users/itsme/projects/
   com.demo.myapp
```

```
Creating a new cordova project.
/Users/itsme/projects/myApp /Users/itsme/projects
Adding android project...
Creating Cordova project for the Android platform:
	Path: platforms/android
	Package: com.demo.myapp
	Name: myApp
	Activity: MainActivity
	Android target: android-23
Android project created with cordova-android@5.0.0
Discovered plugin "cordova-plugin-whitelist" in config.xml. Installing to the project
Fetching plugin "cordova-plugin-whitelist@1" via npm
Installing "cordova-plugin-whitelist" for android

This plugin is only applicable for versions of cordova-android greater than 4.0. If you have a previous platform version, you do *not* need this plugin since the whitelist will be built in.

Fetching plugin "https://github.com/forcedotcom/SalesforceMobileSDK-CordovaPlugin" via git clone
Repository "https://github.com/forcedotcom/SalesforceMobileSDK-CordovaPlugin" checked out to git ref "master".
Installing "com.salesforce" for android
Dependent plugin "cordova-plugin-whitelist" already installed on android.
Fetching plugin "cordova-plugin-device" via npm
Installing "cordova-plugin-device" for android
Fetching plugin "phonegap-plugin-push" via npm
Installing "phonegap-plugin-push" for android
Running SalesforceMobileSDK plugin android post-install script
The node package shelljs is required to use this script. Run 'npm install shelljs' before running this script.
/Users/itsme/projects/myApp

Your application project is ready in /Users/itsme/projects/
```
Then, I ran the following code
```
$ cd myApp
$ cordova build
   BUILD SUCCESSFUL
$ cordova emulate android
   BUILD SUCCESSFUL
   INSTALL SUCCESS
   LAUNCH SUCCESS
```
I see an empty screen in my emulator with below error in the browser console i.e., via `chrome://inspect/#devices`

```
No Content-Security-Policy meta tag found. Please add one when using the cordova-plugin-whitelist plugin. 
```

![screen shot 2015-12-24 at 12 21 18 am](https://cloud.githubusercontent.com/assets/5163972/11982798/6791d4fe-a9d4-11e5-84f8-d9a2ba1e0b77.png)

I added the following line in my `index.html`,
```
<meta http-equiv="Content-Security-Policy" content="default-src *; style-src *; frame-src *;">
```

Checked the `config.html`,

```
<allow-navigation href="https://localhost" />
<allow-navigation href="https://*.force.com" />
<allow-navigation href="https://*.salesforce.com" />
<access origin="https://localhost" />
<access origin="https://*.force.com" />
<access origin="https://*.salesforce.com" />
```
When I re-built it, there was no error or warnings but I still see the empty page. I am not sure why this simple app is not redirecting to the Salesforce login page, am i missing anything?

Thanks in advance.

**Comment by bhariharan** [comment]
The problem is right here (from your console output):

```Running SalesforceMobileSDK plugin android post-install script
The node package shelljs is required to use this script. Run 'npm install shelljs' before running this script.```

Try running ```npm install shelljs``` and re-running the script. Feel free to reopen this issue if you still see problems.

**Comment by aravindtrue** [comment]
I think it is already installed. Tried to check the version and got this,
```
$ sh --version
GNU bash, version 3.2.57(1)-release (x86_64-apple-darwin14)
Copyright (C) 2007 Free Software Foundation, Inc.
```

**Comment by aravindtrue** [comment]
I tried you mentioned in that comment,
```
npm install forcedroid locally in a folder.
npm install shelljs locally in that same folder.
node_modules/.bin/forcedroid create
```
and ended up with below error,
```
Element activity#com.salesforce.androidsdk.phonegap.ui.SalesforceDroidGapActivity at AndroidManifest.xml:30:9-35:20 duplicated with element declared at AndroidManifest.xml:23:9-28:20
/platforms/android/AndroidManifest.xml Error:
Validation failed, exiting
:processDebugManifest FAILED

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':processDebugManifest'.
> Manifest merger failed with multiple errors, see logs
```
I see the activity is created twice with same name,

```
line no: 23

<activity android:configChanges="orientation|keyboardHidden|keyboard|screenSize|locale" android:label="@string/app_name" android:launchMode="singleTop" android:name="com.salesforce.androidsdk.phonegap.ui.SalesforceDroidGapActivity" android:theme="@android:style/Theme.Black.NoTitleBar">
         <intent-filter>
                <action android:name="android.intent.action.MAIN" />
                <category android:name="android.intent.category.LAUNCHER" />
         </intent-filter>
</activity>

line no: 30

<activity android:configChanges="orientation|keyboardHidden|keyboard|screenSize|locale" android:label="@string/app_name" android:name="com.salesforce.androidsdk.phonegap.ui.SalesforceDroidGapActivity" android:theme="@android:style/Theme.Black.NoTitleBar">
         <intent-filter>
                <action android:name="android.intent.action.MAIN" />
                <category android:name="android.intent.category.LAUNCHER" />
         </intent-filter>
</activity>
```

**Comment by aravindtrue** [comment]
Tried,

```
$ cordova platform rm android
$ cordova plugin rm com.salesforce
$ cordova plugin add https://github.com/forcedotcom/SalesforceMobileSDK-CordovaPlugin
   Fetching plugin "https://github.com/forcedotcom/SalesforceMobileSDK-CordovaPlugin" via git clone
   Repository "https://github.com/forcedotcom/SalesforceMobileSDK-CordovaPlugin" checked out to git ref "master".

$ cordova platform add android@5.0.0
  Adding android project...
  Creating Cordova project for the Android platform:
	Path: platforms/android
	Package: com.demo.myapp
	Name: myApp
	Activity: MainActivity
	Android target: android-23
  Android project created with cordova-android@5.0.0
  Discovered plugin "cordova-plugin-whitelist" in config.xml. Installing to the project
  Fetching plugin "cordova-plugin-whitelist@1" via npm
  Installing "cordova-plugin-whitelist" for android

  This plugin is only applicable for versions of cordova-android greater than 4.0. If you have a previous     platform version, you do *not* need this plugin since the whitelist will be built in.
	
  Installing "com.salesforce" for android
  Dependent plugin "cordova-plugin-whitelist" already installed on android.
  Fetching plugin "cordova-plugin-device" via npm
  Installing "cordova-plugin-device" for android
  Fetching plugin "phonegap-plugin-push" via npm
  Installing "phonegap-plugin-push" for android
  Running SalesforceMobileSDK plugin android post-install script
  Fixing application AndroidManifest.xml
  Moving Salesforce libraries to the correct location
  Fixing Gradle dependency paths in Salesforce libraries
  Fixing root level Gradle file for the generated app
  include ":SalesforceSDK"

  include ":SmartStore"

  include ":SmartSync"

  include ":SalesforceHybrid"

  Moving Gradle wrapper files to application directory
  Fixing application build.gradle
  allprojects {
	repositories {
		mavenCentral()
	}
  }
  Done running SalesforceMobileSDK plugin android post-install script
```
Got the below error,

```
/Users/itsme/projects/myApp/platforms/android/build/intermediates/manifests/full/debug/AndroidManifest.xml:35:23-41 : No resource found that matches the given name (at 'icon' with value '@drawable/sf__icon').

:processDebugResources FAILED

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':processDebugResources'.
> com.android.ide.common.process.ProcessException: org.gradle.process.internal.ExecException: Process 'command '/Users/ams/Library/Android/android-sdk-macosx/build-tools/23.0.1/aapt'' finished with non-zero exit value 1
```
Changed `@drawable/sf__icon` to `@drawable/icon` and on re-built, got the same error again. as below,
```
Error:
	Element activity#com.adobe.phonegap.push.PushHandlerActivity at AndroidManifest.xml:30:9-104 duplicated with element declared at AndroidManifest.xml:6:9-135
/Users/itsme/projects/myApp/platforms/android/AndroidManifest.xml Error:
	Validation failed, exiting
:processDebugManifest FAILED

FAILURE: Build failed with an exception.
```

**Comment by bhariharan** [comment]
Looks like manifest merger failed. Did you try removing one of the duplicate entries in the manifest file?

**Comment by aravindtrue** [comment]
Yes, I did but it is created back again.
Looking at http://tools.android.com/tech-docs/new-build-system/user-guide/manifest-merger#TOC-Markers

**Comment by aravindtrue** [comment]
@bhariharan 
So far, no luck. Any workaround?

**Comment by bhariharan** [comment]
@aravindtrue We've fixed these issues in our latest version of ```Mobile SDK 4.1.1```. Grab the latest and give it a shot.

**Comment by PrasantaPardhi** [comment]
I tried with the latest version but still getting a blank "Users" screen. I am trying with hybrid_local. Salesforce user authentication screen is not coming. Any exact steps to do.

**Comment by abhinavguptas** [comment]
Its still coming with latest Cordova and Salesforce SDK, because few Salesforce plugins failed to compile correctly, and thus crashes at runtime during load for "java.lang.ClassNotFoundException: com.salesforce.androidsdk.phonegap.plugin.SalesforceOAuthPlugin"

@bhariharan and Salesforce Mobile SDK team, can you please suggest, when a fix for the same can be expected ?

**Comment by alanlimateles** [comment]
Any sugestion? i am getting problem in manifest too, with duplicate entries.

**Comment by toddhalfpenny** [comment]
I had posted a work around for this before... we have got around this (the java.lang.ClassNotFoundException and duplicate entries in the manifest) by making sure that the SF plugin is the last one to be added. We see this only if we have other plugins included in the project.

**Comment by aravindtrue** [comment]
Oops, I missed this conversation.

@bhariharan 
You're right! The manifest merger issue is fixed and I can build the project without any exception. But, it still shows the blank white screen and getting this error in the console,
```
 Uncaught module com.salesforce.plugin.oauth not found
```

Noticed it while creating a new project,
```
Adding android project...
Creating Cordova project for the Android platform:
	Path: platforms/android
	Package: com.demo.testingapp
	Name: TestingApp
	Activity: MainActivity
	Android target: android-23
Android project created with cordova-android@5.0.0
Discovered plugin "cordova-plugin-whitelist" in config.xml. Installing to the project
Fetching plugin "cordova-plugin-whitelist@1" via npm
Installing "cordova-plugin-whitelist" for android

               This plugin is only applicable for versions of cordova-android greater than 4.0. If you have a previous platform version, you do *not* need this plugin since the whitelist will be built in.
          
Fetching plugin "https://github.com/forcedotcom/SalesforceMobileSDK-CordovaPlugin" via git clone
Repository "https://github.com/forcedotcom/SalesforceMobileSDK-CordovaPlugin" checked out to git ref "v4.1.1".
Installing "com.salesforce" for android
Dependent plugin "cordova-plugin-whitelist" already installed on android.
Fetching plugin "cordova-plugin-device" via npm
Installing "cordova-plugin-device" for android
Fetching plugin "phonegap-plugin-push" via npm
Installing "phonegap-plugin-push" for android
Failed to install 'phonegap-plugin-push':Error: Variable(s) missing: SENDER_ID
    at /usr/local/lib/node_modules/cordova/node_modules/cordova-lib/src/plugman/install.js:358:23
    at _fulfilled (/usr/local/lib/node_modules/cordova/node_modules/q/q.js:787:54)
    at self.promiseDispatch.done (/usr/local/lib/node_modules/cordova/node_modules/q/q.js:816:30)
    at Promise.promise.promiseDispatch (/usr/local/lib/node_modules/cordova/node_modules/q/q.js:749:13)
    at /usr/local/lib/node_modules/cordova/node_modules/q/q.js:557:44
    at flush (/usr/local/lib/node_modules/cordova/node_modules/q/q.js:108:17)
    at _combinedTickCallback (internal/process/next_tick.js:67:7)
    at process._tickCallback (internal/process/next_tick.js:98:9)
Failed to install 'com.salesforce':Error: Variable(s) missing: SENDER_ID
    at /usr/local/lib/node_modules/cordova/node_modules/cordova-lib/src/plugman/install.js:358:23
    at _fulfilled (/usr/local/lib/node_modules/cordova/node_modules/q/q.js:787:54)
    at self.promiseDispatch.done (/usr/local/lib/node_modules/cordova/node_modules/q/q.js:816:30)
    at Promise.promise.promiseDispatch (/usr/local/lib/node_modules/cordova/node_modules/q/q.js:749:13)
    at /usr/local/lib/node_modules/cordova/node_modules/q/q.js:557:44
    at flush (/usr/local/lib/node_modules/cordova/node_modules/q/q.js:108:17)
    at _combinedTickCallback (internal/process/next_tick.js:67:7)
    at process._tickCallback (internal/process/next_tick.js:98:9)
Error: Variable(s) missing: SENDER_ID
/Users/ams

Your application project is ready in /Users/ams/work-mob/.
```
There is a problem while installing `phonegap-plugin-push`.

**Comment by toddhalfpenny** [comment]
Looks like the issue with the push-plugin version (see here http://salesforce.stackexchange.com/questions/113727/sender-id-error-while-creating-hybrid-app-using-forcedroid).
TLDR; version 1.6 needs some extra config, so force an older version (e.g. 1.5)

**Comment by alanlimateles** [comment]
Can you build with cordova build android?

**Comment by kchitalia** [comment]
From what I understand, the push plugin issue should be fixed on our unstable branch. @bhariharan please keep me honest here.

**Comment by bhariharan** [comment]
The push plugin issue is fixed in the latest version of Mobile SDK - ```4.1.2```.
